### PR TITLE
AK: SourceGenerator - Add helpers for outputting functions & namespaces

### PR DESCRIPTION
**Pull request by ant1441**
This introduces two helpers to `SourceGenerator`, `fork_namespace()` and `fork_function()`.

These helpers simplify the C++ code generation, by ensuring that the closing tags are always output,
without needing to remember to include them

### CODE REVIEW

File: AK/SourceGenerator.h [hunk 1]
```diff
@@ -75,10 +82,23 @@ class SourceGenerator {
...
+    void append_include(StringView path)
+    {
+        append(String::formatted("#include <{}>\n", path));
+    }
+
+    SourceGenerator fork_namespace(StringView namespace_)
+    {
+        append(String::formatted("\nnamespace {} {{\n", namespace_));
+        auto suffix = String::formatted("\n}} // namespace {}\n", namespace_);
+        return SourceGenerator { m_builder, m_mapping, m_opening, m_closing, suffix };
```
Comment (kleinesfilmroellchen) [inline AK/SourceGenerator.h:94]: That's a UAF. The SourceGenerator only recieves a StringView, but the string that is referenced is destroyed when the fork_namespace function exits. You probably want to keep the suffix as a string and move() it into the forked generator.

File: AK/SourceGenerator.h [hunk 2]
```diff
@@ -75,10 +82,23 @@ class SourceGenerator {
         }
     }
 
+    void append_include(StringView path)
+    {
+        append(String::formatted("#include <{}>\n", path));
+    }
```
Comment (kleinesfilmroellchen) [inline AK/SourceGenerator.h:88]: This can be in its own commit.

File: AK/SourceGenerator.h [hunk 3]
```diff
@@ -94,6 +94,13 @@ class SourceGenerator {
         return SourceGenerator { m_builder, m_mapping, m_opening, m_closing, suffix };
     }
 
+    SourceGenerator fork_function(StringView return_type, StringView name, StringView arguments)
```
Comment (kleinesfilmroellchen) [inline AK/SourceGenerator.h:97]: With this signature we can't use the method to generate const functions. Let's have another parameter `Constness` (I hope you're familiar with the single-purpose bool-like enums). Also, let's rename return_type to something more generic, as that string can also be used for attributes like ALWAYS_INLINE.

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]
I'd like to say that at this point it should be fairly easy to have an indentation level stored in the generator and increase the level for certain kinds of forks, like fork_function. That should make the generated code less susceptible to misindents.

File: AK/SourceGenerator.h (hunk 1 above)
Comment (ant1441) [reply AK/SourceGenerator.h:94]: I spent a good while debugging that exact bug, when I had the `m_suffix` member as `StringView`.
I guess it's currently working without the move() by luck.

I'm more familiar with Rust's default move, so I'm definitely less familiar with when C++ needs move().
Thanks :+1:

Comment (kleinesfilmroellchen) [reply AK/SourceGenerator.h:94]: The point here is not move(), the point is that with StringView you don't own the string data. In Rust terms: it's a `&str` where the lifetime of the borrowed data has expired.
